src/engine/Operation.cpp

```diff
@@ -211,144 +209,95 @@ void Operation::checkTimeout() const {
-    size_t timeInMilliseconds) {
-  updateRuntimeInformationOnSuccess(
-      *resultAndCacheStatus._resultPointer->resultTable(),
-      resultAndCacheStatus._cacheStatus, timeInMilliseconds,
-      resultAndCacheStatus._resultPointer->runtimeInfo());
-}
-
 // _____________________________________________________________________________
 void Operation::updateRuntimeInformationWhenOptimizedOut(
-    std::vector<RuntimeInformation> children,
     RuntimeInformation::Status status) {
-  _runtimeInfo.status_ = status;
-  _runtimeInfo.children_ = std::move(children);
+  _runtimeInfo->status_ = status;
   // This operation was optimized out, so its operation time is zero.
   // The operation time is computed as
   // `totalTime_ - #sum of childrens' total time#` in `getOperationTime()`.
   // To set it to zero we thus have to set the `totalTime_` to that sum.
-  auto timesOfChildren = _runtimeInfo.children_ |
+  auto timesOfChildren = _runtimeInfo->children_ |
                          std::views::transform(&RuntimeInformation::totalTime_);
-  _runtimeInfo.totalTime_ =
+  _runtimeInfo->totalTime_ =
       std::accumulate(timesOfChildren.begin(), timesOfChildren.end(), 0.0);
 
   if (_executionContext) {
-    _executionContext->signalQueryUpdate(_runtimeInfo);
-  }
-}
-
-// _____________________________________________________________________________
-void Operation::updateRuntimeInformationWhenOptimizedOut(
-    RuntimeInformation::Status status) {
-  auto setStatus = [&status](RuntimeInformation& rti,
-                             const auto& self) -> void {
-    rti.status_ = status;
-    rti.totalTime_ = 0;
-    for (auto& child : rti.children_) {
-      self(child, self);
-    }
-  };
-  setStatus(_runtimeInfo, setStatus);
-
-  if (_executionContext) {
-    _executionContext->signalQueryUpdate(_runtimeInfo);
+    _executionContext->signalQueryUpdate(*_rootRuntimeInfo);
   }
 }
 
 // _______________________________________________________________________
 void Operation::updateRuntimeInformationOnFailure(size_t timeInMilliseconds) {
-  _runtimeInfo.children_.clear();
-  for (auto child : getChildren()) {
-    _runtimeInfo.children_.push_back(child->getRootOperation()->_runtimeInfo);
-  }
-
-  _runtimeInfo.totalTime_ = timeInMilliseconds;
-  _runtimeInfo.status_ = RuntimeInformation::Status::failed;
+  _runtimeInfo->totalTime_ = timeInMilliseconds;
+  _runtimeInfo->status_ = RuntimeInformation::Status::failed;
 
   if (_executionContext) {
-    _executionContext->signalQueryUpdate(_runtimeInfo);
+    _executionContext->signalQueryUpdate(*_rootRuntimeInfo);
   }
 }
 
 // __________________________________________________________________
-void Operation::createRuntimeInfoFromEstimates() {
-  _runtimeInfo.setColumnNames(getInternallyVisibleVariableColumns());
+void Operation::createRuntimeInfoFromEstimates(
+    std::shared_ptr<RuntimeInformation> runtimeInformation,
+    std::shared_ptr<RuntimeInformation> rootRuntimeInformation) {
```

Comment:
I think the more idiomatic pattern would be:
* Each Operation initially has a valid `shared_ptr` (make_shared with default constructor, basically as it was).
* Then `createRuntimInfoFromEstimates` consistently gets one argument, namely the pointer to the root.
* You then have garbage (empty) runtime infos in the tests that don't care for them, but that is not an issue (as it was before).
* Then for most of the cases you just have to change `.` to `->`, or better to `runtimeInfo().`
* The only things were have to manually do stuff is the read from cache case (reset the children pointers!), and optimized out case (depends on the exact call). Maybe we should have a talk about this.